service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -103,34 +73,59 @@ private void createNotExistingTag(Set<Tag> tags) {
     }
 
     @Override
-    public boolean remove(Long id) {
-        if (!certificateRepository.getById(id).isPresent()) {
-            throw new EntityNotFoundException(" (id = " + id + ")");
+    public CertificateDTO update(Long id, CertificateDTO certificateDTO) {
+        CertificateDTO formerCertificate = getById(id);
+        String newName = certificateDTO.getName();
+        if (certificateRepository.getByName(newName).isPresent()
+                && !newName.equals(formerCertificate.getName())) {
+            throw new EntityAlreadyExistsException(" (name = " + newName + ")");
         }
-        boolean result = certificateRepository.deleteById(id);
-        if (!result) {
-            throw new DeleteEntityException(" (id = " + id + ")");
+        Certificate certificate = converter.convertToEntity(certificateDTO);
+        Set<Tag> updatedTags = certificate.getTags();
+        if (updatedTags != null) {
+            createNotExistingTag(updatedTags);
         }
-        return result;
+        certificate.setId(id);
+        certificate.setCreateDate(formerCertificate.getCreateDate());
+        certificate.setLastUpdateDate(LocalDateTime.now());
+        return converter.convertToDto(certificateRepository.update(certificate));
     }
 
-    @Override
-    public List<CertificateDTO> executeQuery(CertificatePageQueryDTO queryDTO) {
-        SqlSpecification specification = new CertificateAllSpecification(queryDTO.getTagName(),
-                queryDTO.getContext(), queryDTO.getSortBy(), queryDTO.getOrder());
-        List<Certificate> certificates = certificateRepository.query(specification);
-        addTagsToListCertificates(certificates);
-        return converter.convertToListDTO(certificates);
-    }
 
-    private void addTagsToListCertificates(List<Certificate> certificates) {
-        certificates.forEach(this::addTagsToCertificate);
+    @Override
+    public PageDTO<CertificateDTO> findByParams(CertificatePageQueryDTO queryDTO, PageRequestDTO pageRequestDTO) {
+        CriteriaSpecification<Certificate> specification = new CertificateByParamsSpecification(queryDTO.getTags(),
```

Comment:
`new CertificateByParamsSpecification(queryDTO.getTags(),queryDTO.getName(), queryDTO.getDescription(), queryDTO.getSortBy(), queryDTO.getOrder())`
replace to
`new CertificateByParamsSpecification(queryDTO)`
??? is it possible?

Reply:
the problem is that the CertificatePageQueryDTO.class is at the service level and is not visible in the persistence module